includes/class-signature.php

```diff
@@ -107,12 +124,244 @@ public static function generate_signature( $user_id, $http_method, $url, $date,
+		} else {
+			$request = self::format_server_request( $request );
+			$headers = $request['headers']; // $_SERVER array
+			$headers['(request-target)'][0] = strtolower( $headers['request_method'][0] ) . ' ' . $headers['request_uri'][0];
+		}
+
+		if ( ! isset( $headers['signature'] ) ) {
+			return new \WP_Error( 'activitypub_signature', 'Request not signed', array( 'status' => 403 ) );
+		}
+
+		if ( array_key_exists( 'signature', $headers ) ) {
+			$signature_block = self::parse_signature_header( $headers['signature'] );
+		} elseif ( array_key_exists( 'authorization', $headers ) ) {
+			$signature_block = self::parse_signature_header( $headers['authorization'] );
+		}
+
+		if ( ! isset( $signature_block ) || ! $signature_block ) {
+			return new \WP_Error( 'activitypub_signature', 'Incompatible request signature. keyId and signature are required', array( 'status' => 403 ) );
+		}
+
+		$signed_headers = $signature_block['headers'];
+		if ( ! $signed_headers ) {
+			$signed_headers = array( 'date' );
+		}
+
+		$signed_data = self::get_signed_data( $signed_headers, $signature_block, $headers );
+		if ( ! $signed_data ) {
+			return new \WP_Error( 'activitypub_signature', 'Signed request date outside acceptable time window', array( 'status' => 403 ) );
+		}
+
+		$algorithm = self::get_signature_algorithm( $signature_block );
+		if ( ! $algorithm ) {
+			return new \WP_Error( 'activitypub_signature', 'Unsupported signature algorithm (only rsa-sha256 and hs2019 are supported)', array( 'status' => 403 ) );
+		}
+
+		if ( \in_array( 'digest', $signed_headers, true ) && isset( $body ) ) {
+			if ( is_array( $headers['digest'] ) ) {
+				$headers['digest'] = $headers['digest'][0];
+			}
+			$digest = explode( '=', $headers['digest'], 2 );
+			if ( 'SHA-256' === $digest[0] ) {
+				$hashalg = 'sha256';
+			}
+			if ( 'SHA-512' === $digest[0] ) {
+				$hashalg = 'sha512';
+			}
+
+			if ( \base64_encode( \hash( $hashalg, $body, true ) ) !== $digest[1] ) { // phpcs:ignore
+				return new \WP_Error( 'activitypub_signature', 'Invalid Digest header', array( 'status' => 403 ) );
+			}
+		}
+
+		$public_key = self::get_remote_key( $signature_block['keyId'] );
+		if ( \is_wp_error( $public_key ) ) {
+			return $public_key;
+		}
+		$verified = \openssl_verify( $signed_data, $signature_block['signature'], $public_key, $algorithm ) > 0;
+		if ( ! $verified ) {
+			return new \WP_Error( 'activitypub_signature', 'Invalid signature', array( 'status' => 403 ) );
+		}
+		return $verified;
+	}
+
+	/**
+	 * Get public key from key_id
+	 *
+	 * @param string $key_id
+	 * @return string $publicKeyPem
+	 * @author Django Doucet <django.doucet@webdevstudios.com>
+	 */
+	public static function get_remote_key( $key_id ) { // phpcs:ignore
+		$actor = \Activitypub\get_remote_metadata_by_actor( strtok( strip_fragment_from_url( $key_id ), '?' ) ); // phpcs:ignore
```

Comment:
the meta is now also stored int the followers collection. you do not have to get the remote data any more.

Reply:
Sometimes we need to get a key from a non-follower, ex: someone commenting on our post.

Reply:
ok, true!